### ISSUE TITLE
Implement Density Function Functions

### ISSUE THREAD

**Issue by PersonTheCat** [PersonTheCat/pangaea#27]
# Density Function Functions

This will be an experiment (since XJS will give us _real_ functions), but I want to experiment with the possibility of giving users a way to dynamically create density function codecs in a JSON-friendly way. Here's how that will look:

## Design Example

in  `data/my_mod/pg/function/noodle_noise.json`
```json
{
  "type": "range_choice",
  "input": "y",
  "max_exclusive": {
    "type": "pg:argument",
    "name": "max_height"
  },
  "min_inclusive": {
    "type": "pg:argument",
    "name": "min_height"
  }
  "when_in_range": {
    "type": "noise",
    "noise": "noodle",
    "xz_scale": {
      "type": "pg:argument",
      "name": "xz_scale",
      "default": 1.0
    }
    "y_scale": {
      "type": "pg:argument",
      "name": "z_scale",
      "default": 1.0
    }
  },
  "when_out_of_range": -1.0
}
```

## Explanation

This produces a regular density function, wrapped in a `DensityFunctionFunction`, which also implements `DensityFunction`. 

Here's how it works:
* Creation of this object immediately registers a map codec, the ID of which is the same as the ID of this file. 
* The generated codec accepts an arbitrary map, being the parameters defined here.
* Each parameter is just a regular density function
* When `DensityFunction#mapAll` is called on a `DensityFunctionFunction`, it scans for arguments and inlines them with values from the map

## Design with Structural Codec

Here's a second example using a hypothetical structural codec to produce arguments:
```json
{
  "type": "range_choice",
  "input": "y",
  "max_exclusive": { "arg": "max_height" },
  "min_inclusive": { "arg": "min_height" },
  "when_in_range": {
    "type": "noise",
    "noise": "noodle",
    "xz_scale": { "arg": "xz_scale", "default": 1.0 }
    "y_scale": { "arg": "y_scale", "default": 1.0 }
  },
  "when_out_of_range": -1.0
}
```

## Using the Generated Codec

Finally, here's an example of how this function would be used:
```json
{
  "type": "my_mod:noodle_noise",
  "max_height": 321,
  "min_height": -60
}
```

## Likely Hiccups

The first issue we're likely to encounter with this design is that it requires registering a new codec to `BuiltInRegistries#DENSITY_FUNCTION_TYPE`. We may attempt to unfreeze the registry and fail. 

The second issue is that we're simultaneously designing a _much_ better solution in the form of real functions via XJS. For example,

in `example.xjs`:
```js
// ignoring the types and or documentation, either of which is technically required
fn noodleNoise(minHeight, maxHeight, xzScale = 1.0, yScale = 1.0) {{
  type: 'range_choice',
  input: 'y',
  max_exclusive: maxHeight,
  min_inclusive: minHeight,
  when_in_range: {
    type: 'noise',
    noise: 'noodle',
    xz_scale: xzScale,
    y_scale: yScale,
  },
  when_out_of_range: -1.0,
}}
``` 

in caller file:
```js
import { noodleNoise } from example

noodleNoise {
  maxHeight: 321,
  minHeight: -60,
}
```

**Comment by PersonTheCat** [comment]
The previous design will not work because the argument values in the example are regular numbers, not density functions. For this to work, we may have to operate directly on the JSON data. I don't like this, but here's how it works:
* Creation of this object immediately registers a map codec, as before
* The generated codec accepts an arbitrary map of `string -> DensityFunction`, as before. 
* The original data is stored as plain JSON data, or another common format
* To produce a Density Function, the data must be converted to the common format, then inlined, and finally passed through the regular density function codec

This is quite a nasty solution. I'll put some more thought into how we could revise this, but it's definitely looking like real XJS functions are the way to go

**Comment by PersonTheCat** [comment]
This was not too bad, but it did require an additional Mixin on the Fabric platform to manually register the new registry as a dynamic registry so that entries get prefixed correctly. 

There is, however, yet another problem with this approach: once we figure out how to unfreeze the registry and insert new entries, we would also have to figure out how to remove the dynamic entries each time the server shuts down. This is _really_ nasty...

An alternate approach, yet again, would be to create new registry to house the dynamically-generated codecs. Then, we would again modify the density function codec to have it search this registry for any dynamic codecs before / after checking the main registry. This is tricky, because the dynamic registries aren't available right away. If we try to create a by-name codec from a dynamic registry, it isn't guaranteed to always be current. This means we either need to update CatLib to support that _or_ some kind of codec supplier that acquires the codec each time it's used... 

Long story short: this is _way_ too much effort... and it's very, _very_ sloppy.

**Comment by PersonTheCat** [comment]
There is a final approach I'm willing to try: instead figuring out how to register this new codec dynamically, we can just modify `StructuralDensityCodec` to invoke the generated functions with a special syntax. For example, we could require the user to provide an `invoke` or `template` key instead of a `type` key. The structural codec can get the current template from `PgRegistries#DENSITY_TEMLPATE`. 

Here's how that would look:
```json
{
  "template": "my_mod:noodle_noise",
  "max_height": 321,
  "min_height": -60
}
```

The only problem with _this_ solution is that it doesn't scale easily. We would need some way to conveniently wrap any given codec to provide this feature.

Once again, I _don't_ love it... but I'll give it a try since I'm still enjoying this for now.

**Comment by PersonTheCat** [comment]
I wound up implementing this feature in a way that does support using the `type` field. It's quite a hack and I will most likely will get rid of it at some point before launch, but I will be experimenting with it for now and looking for additional feedback.
